src/test/documentation_examples.rs

```diff
@@ -0,0 +1,1385 @@
+use futures::TryStreamExt;
+
+use crate::{
+    bson::{doc, Bson},
+    error::Result,
+    options::FindOptions,
+    test::TestClient,
+    Collection,
+};
+
+macro_rules! assert_coll_count {
+    ($coll:expr, $expected:expr) => {
+        assert_eq!($coll.count_documents(None, None).await.unwrap(), $expected);
+    };
+}
+
+macro_rules! assert_cursor_count {
+    ($cursor:expr, $expected:expr) => {{
+        let docs: Vec<_> = $cursor.try_collect().await.unwrap();
+        assert_eq!(docs.len(), $expected);
+    }};
+}
+
+macro_rules! run_on_each_doc {
+    ($cursor:expr, $name:ident, $check:block) => {{
+        let mut cursor = $cursor;
+
+        while let Some($name) = cursor.try_next().await.unwrap() $check;
```

Comment:
I originally wrote this to take lambda as an expression rather than a name and a block and then called the lambda inside the loop, but it sometimes ran into type inference issues on the type of the parameter, and I couldn't figure out why it was only happening sometimes. Using a name and then a block doesn't seem to cause these issues, and it ends up generating code closer to what I'd actually write without a macro, so I figured this was fine.